baker_cli.go

```diff
@@ -0,0 +1,181 @@
+package baker
+
+import (
+	"flag"
+	"fmt"
+	"html/template"
+	"math/rand"
+	"net"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	flagHelpConfig = flag.String("help", "", "show help for a `component` (input/filter/output/upload) (use '*' to dump all)")
+	flagVersion    = flag.Bool("version", false, "print build version number")
+	flagVerbose    = flag.Bool("v", false, "verbose logging (debug level)")
+	flagQuiet      = flag.Bool("q", false, "quiet logging (warn level)")
+	flagPretty     = flag.Bool("pretty", false, "human-readable logging (unstructured logging)")
+	flagPProf      = flag.String("pprof", "", `run pprof server on host port provided (disabled if ""), use "localhost:"  for a free port`)
+)
+
+// Use `-ldflags="-X 'github.com/AdRoll/baker.BuildVersion=someversion'"` when building Baker to set this value
+var BuildVersion = "-- unknown --"
+
+// MainCLI provides a handy way to quickly create a Baker program only providing
+// the available Components.
+//
+// The function includes many utilities that can be configured by command line arguments:
+//  -help: Prints available options and components
+//  -version: print build version (build with `-ldflags="-X 'github.com/AdRoll/baker.BuildVersion=someversion'"` to set the value)
```

Comment:
not sure we should include the version flag here. this is a more advanced use, requiring to build baker with specific flag.
Note that flags could also been declared by whatever main application calling MainCLI

Reply:
how would you show the version number? if the `main()` declares a flag, it can't use it until `flag.Parse()` is called in `mainCLI`, but once called then the control is in the hand of `mainCLI` and the original `main` fn can't add options for the flags it declared (like printing the version and exit).

Reply:
maybe one idea, though a bit cumbersome (or cucumbersome :laughing: ) could be to define some functions in the `main` and pass them to `MainCLI`, that calls them after parse:

```go
// main()
var fns = []func(){
	func() {
		if *flagVersion {
			fmt.Printf("Baker version: %s\n", build)
			os.Exit(0)
		}
	},
}
baker.MainCLI(components, fns)

// MainCLI
func MainCLI(components Components, closures []func()) error {
	flag.Parse()
	for _, fn := range closures {
		fn()
	}
}
```

Reply:
I haven't thought about that when I wrote my first comment.
I'm not crazy about passing additional functions. 
I think MainCLi should be enough for most cases, if one needs more, they can just copy baker.MainCLI into their code and modify it to suit their needs, seems reasonable enough